From 9e0c773761bf908e671c1000d768833290508d6d Mon Sep 17 00:00:00 2001
From: Ankit Agrawal <ankit.agarwal@intel.com>
Date: Tue, 19 Dec 2023 09:59:58 +0530
Subject: [PATCH] Fixed for Media crash during Monkey Test Run.

Observed following crash
java.lang.NullPointerException Attempt to invoke virtual method
'java.lang.Object android.car.Car.getCarManager(java.lang.String)' on a
null object reference.

Creating Car object based on car service listener. Once car service is
present, it will get connected and get the car object.

Tests:
Run monkey test and observe, test is pass.

Tracked-On: OAM-114265
Signed-off-by: Ankit Agrawal <ankit.agarwal@intel.com>
---
 src/com/android/car/media/MediaActivity.java | 20 +++++++++++++++++---
 1 file changed, 17 insertions(+), 3 deletions(-)

diff --git a/src/com/android/car/media/MediaActivity.java b/src/com/android/car/media/MediaActivity.java
index 591285a..ac6e1a1 100644
--- a/src/com/android/car/media/MediaActivity.java
+++ b/src/com/android/car/media/MediaActivity.java
@@ -25,6 +25,7 @@ import android.app.AlertDialog;
 import android.app.Application;
 import android.app.PendingIntent;
 import android.car.Car;
+import android.car.Car.CarServiceLifecycleListener;
 import android.car.content.pm.CarPackageManager;
 import android.car.drivingstate.CarUxRestrictions;
 import android.car.media.CarMediaIntents;
@@ -129,6 +130,17 @@ public class MediaActivity extends FragmentActivity implements MediaActivityCont
     }
 
 
+    private CarServiceLifecycleListener mCarServiceLifecycleListener = (car, ready) -> {
+        if (!ready) {
+            Log.d(TAG, "Disconnect from Car Service");
+            return;
+        }
+        Log.d(TAG, "Connected to Car Service");
+        mCar = car; 
+        mCarPackageManager = (CarPackageManager) mCar.getCarManager(Car.PACKAGE_SERVICE);
+    };
+
+
     @Override
     protected void onCreate(Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
@@ -181,8 +193,8 @@ public class MediaActivity extends FragmentActivity implements MediaActivityCont
         playbackViewModel.getPlaybackStateWrapper().observe(this,
                 state -> handlePlaybackState(state, true));
 
-        mCar = Car.createCar(this);
-        mCarPackageManager = (CarPackageManager) mCar.getCarManager(Car.PACKAGE_SERVICE);
+        Car.createCar(this, /* handler= */ null,
+                Car.CAR_WAIT_TIMEOUT_WAIT_FOREVER, mCarServiceLifecycleListener);
 
         mMediaActivityController = new MediaActivityController(this, getMediaItemsRepository(),
                 mCarPackageManager, mBrowseContainer);
@@ -230,7 +242,9 @@ public class MediaActivity extends FragmentActivity implements MediaActivityCont
 
     @Override
     protected void onDestroy() {
-        mCar.disconnect();
+        if (mCar != null) {
+            mCar.disconnect();
+        }
         mMediaActivityController.onDestroy();
         super.onDestroy();
     }
-- 
2.17.1

